File: kotlin/src/main/kotlin/com/automation/agent/agent/covfixer/Analyze.kt
package com.automation.agent.agent.covfixer

import com.automation.agent.agent.fixflow.AnalyzeInput
import com.automation.agent.agent.fixflow.FileEdit
import com.automation.agent.agent.fixflow.FileWork
import com.automation.agent.agent.fixflow.explore
import com.automation.agent.agent.fixflow.extractJsonArray
import com.automation.agent.agent.fixflow.parallelAnalyze
import com.automation.agent.agent.fixflow.readFile
import com.automation.agent.agent.fixflow.stripFences
import com.automation.agent.agent.setup.generateText
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private val planJson = Json { ignoreUnknownKeys = true }

private val log: System.Logger = System.getLogger("automation-agent.covfixer")

/**
 * Plans test placement by having a tool-using agent examine the checked-out repo's real conventions,
 * then generates a test per file in parallel from that plan.
 */
suspend fun analyze(input: AnalyzeInput): List<FileEdit> {
    val plan = explorePlan(input)
    return execute(input, plan)
}

/** The explorer's decision for one source file, grounded in the repo's actual existing tests. */
@Serializable
internal data class PlanEntry(
    @SerialName("source") val source: String = "",
    @SerialName("test_path") val testPath: String = "",
    @SerialName("framework") val framework: String = "",
    @SerialName("notes") val notes: String = "",
)

/**
 * Runs a tool-using agent that navigates the checkout itself (read_file / list_dir) to learn the
 * repo's real test conventions and returns a per-file plan keyed by source path.
 */
private suspend fun explorePlan(input: AnalyzeInput): Map<String, PlanEntry> {
    val model = requireNotNull(input.llm) { "explore: an LLM is required" }
    val out = explore(model, input.repoDir, prompts.get("explore"), buildExploreInput(input.work))
    val plan = parsePlan(out)
    require(plan.isNotEmpty()) { "explore: produced no test placements" }
    return plan
}

/** Writes each test from the plan + source, one parallel agent per file. */
private suspend fun execute(input: AnalyzeInput, plan: Map<String, PlanEntry>): List<FileEdit> =
    parallelAnalyze(input.work) { w ->
        val entry = plan[w.path]
        val src = if (entry == null || entry.testPath.isBlank()) null else runCatching { readFile(input.repoDir, w.path) }.getOrNull()
        if (entry == null || src == null) {
            val reason = if (entry == null) "no test placement from explorer" else "unreadable source"
            log.log(System.Logger.Level.WARNING, "coverage analyze: skipping ${w.path} ($reason)")
            FileEdit("", "") // explorer couldn't place it, or unreadable -> skip
        } else {
            val model = requireNotNull(input.coder()) { "execute: an LLM is required" }
            val out = generateText(model, prompts.get("analyze"), buildExecuteInput(w, src, entry, input.feedback))
            FileEdit(path = entry.testPath, content = stripFences(out))
        }
    }

internal fun parsePlan(out: String): Map<String, PlanEntry> {
    val js = extractJsonArray(out)
    require(js.isNotEmpty()) { "no JSON array in explorer output" }
    val entries =
        try {
            planJson.decodeFromString<List<PlanEntry>>(js)
        } catch (e: Exception) {
            throw IllegalArgumentException("decode plan JSON: ${e.message}")
        }
    return entries.filter { it.source.isNotBlank() }.associateBy { it.source }
}

internal fun buildExploreInput(work: List<FileWork>): String =
    buildString {
        append("Source files that need tests:\n")
        work.forEach { append("- ${it.path}\n") }
    }

internal fun buildExecuteInput(w: FileWork, src: String, p: PlanEntry, ciFeedback: String): String =
    buildString {
        append("Write the test file at: ${p.testPath}\nFramework / convention: ${p.framework}\n")
        if (p.notes.isNotBlank()) append("Notes: ${p.notes}\n")
        append("\nUncovered logic to cover:\n")
        w.items.forEach { append("- $it\n") }
        if (ciFeedback.isNotEmpty()) {
            append("\nThe previous attempt failed CI with:\n")
            append(ciFeedback)
            append("\n")
        }
        append("\nSource file (${w.path}):\n```\n")
        append(src)
        append("\n```\n")
    }

File: go/internal/agent/covfixer/analyze.go
package covfixer

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"automation-agent/internal/agent/fixflow"
	"automation-agent/internal/agent/setup"
)

// Analyze plans test placement by having a tool-using agent examine the checked-out
// repo's real conventions, then generates a test per file in parallel from that plan.
func Analyze(ctx context.Context, in fixflow.AnalyzeInput) ([]fixflow.FileEdit, error) {
	plan, err := explore(ctx, in)
	if err != nil {
		return nil, err
	}
	return execute(ctx, in, plan)
}

// planEntry is the explorer's decision for one source file, grounded in the repo's
// actual existing tests (never derived from a fixed rule).
type planEntry struct {
	Source    string `json:"source"`
	TestPath  string `json:"test_path"`
	Framework string `json:"framework"`
	Notes     string `json:"notes"`
}

// explore runs a tool-using agent that navigates the checkout itself (read_file /
// list_dir) to learn the repo's real test conventions and returns a per-file plan.
func explore(ctx context.Context, in fixflow.AnalyzeInput) (map[string]planEntry, error) {
	out, err := fixflow.Explore(ctx, in.LLM, in.RepoDir, prompts.MustGet("explore"), buildExploreInput(in.Work))
	if err != nil {
		return nil, fmt.Errorf("explore: %w", err)
	}
	plan, err := parsePlan(out)
	if err != nil {
		return nil, fmt.Errorf("explore: %w", err)
	}
	if len(plan) == 0 {
		return nil, fmt.Errorf("explore: produced no test placements")
	}
	return plan, nil
}

// execute writes each test from the plan + source, one parallel agent per file.
func execute(ctx context.Context, in fixflow.AnalyzeInput, plan map[string]planEntry) ([]fixflow.FileEdit, error) {
	return fixflow.ParallelAnalyze(ctx, in.Work, func(ctx context.Context, w fixflow.FileWork) (fixflow.FileEdit, error) {
		p, ok := plan[w.Path]
		if !ok || strings.TrimSpace(p.TestPath) == "" {
			// Explorer couldn't place a test for this file -> skip. Log it so a skip is
			// distinguishable from "nothing to do".
			in.Logger().Warn("coverage analyze: skipping file with no test placement", "path", w.Path)
			return fixflow.FileEdit{}, nil
		}
		src, err := fixflow.ReadFile(in.RepoDir, w.Path)
		if err != nil {
			in.Logger().Warn("coverage analyze: skipping unreadable file", "path", w.Path, "err", err)
			return fixflow.FileEdit{}, nil
		}
		out, err := setup.GenerateText(ctx, in.Coder(), prompts.MustGet("analyze"), buildExecuteInput(w, src, p, in.Feedback))
		if err != nil {
			return fixflow.FileEdit{}, err
		}
		return fixflow.FileEdit{Path: p.TestPath, Content: fixflow.StripFences(out)}, nil
	})
}

func parsePlan(out string) (map[string]planEntry, error) {
	js := fixflow.ExtractJSONArray(out)
	if js == "" {
		return nil, fmt.Errorf("no JSON array in explorer output")
	}
	var entries []planEntry
	if err := json.Unmarshal([]byte(js), &entries); err != nil {
		return nil, fmt.Errorf("decode plan JSON: %w", err)
	}
	m := make(map[string]planEntry, len(entries))
	for _, e := range entries {
		if strings.TrimSpace(e.Source) != "" {
			m[e.Source] = e
		}
	}
	return m, nil
}

func buildExploreInput(work []fixflow.FileWork) string {
	var b strings.Builder
	b.WriteString("Source files that need tests:\n")
	for _, w := range work {
		fmt.Fprintf(&b, "- %s\n", w.Path)
	}
	return b.String()
}

func buildExecuteInput(w fixflow.FileWork, src string, p planEntry, ciFeedback string) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Write the test file at: %s\nFramework / convention: %s\n", p.TestPath, p.Framework)
	if strings.TrimSpace(p.Notes) != "" {
		fmt.Fprintf(&b, "Notes: %s\n", p.Notes)
	}
	b.WriteString("\nUncovered logic to cover:\n")
	for _, u := range w.Items {
		fmt.Fprintf(&b, "- %s\n", u)
	}
	if ciFeedback != "" {
		b.WriteString("\nThe previous attempt failed CI with:\n")
		b.WriteString(ciFeedback)
		b.WriteString("\n")
	}
	fmt.Fprintf(&b, "\nSource file (%s):\n```\n%s\n```\n", w.Path, src)
	return b.String()
}
